src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp

```diff
@@ -2,9 +2,22 @@
 #include <bsp/BSPContainer.h>
 
 InterlocStateHandler::InterlocStateHandler() :
-    m_currentState(&InterlocStateContainer::getState(InterlocStates::DEFAULT)) {}
+    m_currentStateName(InterlocStates::DEFAULT),
+    m_currentState(&InterlocStateContainer::getState(InterlocStates::DEFAULT)),
+    m_stateTracer(m_stateTracerData.data(), MAX_TRACER_TRANSITIONS) {}
 
-void InterlocStateHandler::setState(InterlocStates state) {
+void InterlocStateHandler::setState(InterlocStates state, InterlocEvent event) {
+    if (m_stateTracer.isFull()) {
+        m_stateTracer.pop();
+    }
+
+    auto& stateTransition = m_stateTracer.getNextAllocation().value().get();
+    stateTransition.m_fromState = m_currentStateName;
+    stateTransition.m_toState = state;
+    stateTransition.m_transitionEvent = event;
+    m_stateTracer.push(stateTransition);
```

Comment:
You push the same value you just got from the queue? Won't this just invariably overwrite the last value?

Reply:
The getNextAllocation() returns a reference to the next empty slot in the queue, so I change the data in that slot. The push then just changes the counters in the queue. The push uses a reference and doesn't copy the data, so if I didn't use the getNextAllocation() first, the data I push would go out of scope the moment I exit the function.

Reply:
Oh okay I see.

Reply:
The push is supposed to copy the data though, as of now you are copying twice and not calling the destructor when overwriting it (which is not an issue in this case)

Reply:
getNextAllocation should only be used with advance afterwards. push == getNextAllocation + advance. It's just if you want the  variable temporarly and the confirm if you advance (or not), you can use getNextAllocation